main/src/main/scala/sbt/Defaults.scala

```diff
@@ -601,19 +601,18 @@ object Defaults extends BuildCommon {
         (art, file) <- m.artifacts if art.`type` == Artifact.DefaultType
       } yield file
     def file(id: String) = files(id).headOption getOrElse sys.error(s"Missing ${id}.jar")
-    val allFiles = toolReport.modules.flatMap(_.artifacts.map(_._2))
+    val allJars = toolReport.modules.flatMap(_.artifacts.map(_._2))
     val libraryJar = file(ScalaArtifacts.LibraryID)
     val binVersion = scalaBinaryVersion.value
     val compilerJar =
       if (ScalaInstance.isDotty(scalaVersion.value))
         file(ScalaArtifacts.dottyID(binVersion))
       else file(ScalaArtifacts.CompilerID)
-    val otherJars = allFiles.filterNot(x => x == libraryJar || x == compilerJar)
     new ScalaInstance(scalaVersion.value,
-                      makeClassLoader(state.value)(libraryJar :: compilerJar :: otherJars.toList),
+                      makeClassLoader(state.value)(allJars.toList),
```

Comment:
interesting, but doesn't this mean the scalac jar isn't on the classpath?

I once used a monkey patch to allow custom classes to go ahead of the scala-compiler jar https://gitlab.com/fommil/attic/blob/master/sbt-scala-monkey/src/main/scala/ScalaMonkeyPlugin.scala#L60

Reply:
oh, I stand corrected... I was also filtering out the compiler / library jars.